File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/UpgradeShineParticleEffect.ts
import { FairyGUI } from "csharp";
import {AbstractEffect} from "mods/ModTheSpire/Scripts/Effect/AbstractEffect";
import UIHelper from "mods/ModTheSpire/Scripts/UI/UIHelper";
import {Mathf} from "Core/Module/Math/Mathf";
import Color from "mods/ModTheSpire/Scripts/DataDefine/Color";
import {TimeKit} from "Core/Utils/TimeKit";

export class UpgradeShineParticleEffect extends AbstractEffect{
    public Loader:FairyGUI.GLoader[] = new FairyGUI.GLoader[2];
    private _Interval:number;
    private _Scale:number;
    private _Vx:number;
    private _Vy:number;
    private _Floor:number;
    private static _Gravity = 3000;
    public constructor(private _X,private _Y) {
        super();
        for (let i = 0; i < 2; i++) {
            this.Create(i);
        }
    }
    
    private Create(i:number){
        this.Loader[i] = UIHelper.CreateGLoader();
        this._Interval = Mathf.RandomRange(0.5,1);
        this.Loader[i].SetPivot(0.5,0.5,true);
        this.Loader[i].color = new Color(Mathf.RandomRange(127,255),Mathf.RandomRange(102, 153),Mathf.RandomRange(0, 51),255).UnityColor();
        if(this._Interval > 0.9)
            this.Loader[i].alpha = 0;
        else
            this.Loader[i].alpha = 1;
        this.Loader[i].rotation = Mathf.RandomRange(0,360);
        this._Vx = Mathf.RandomRange(-900,900);
        this._Vx = Mathf.RandomRange(-500,1000);
        this._Floor = Mathf.RandomRange(100,250);
        this.Loader[i].blendMode = FairyGUI.BlendMode.Add;
        this.Loader[i].SetPosition(this._X,this._Y,0);
        if(this._Interval > 0.9)
            this.Loader[i].TweenFade(1,Mathf.Clamp(this._Interval - 0.9,0,0.1)).SetEase(FairyGUI.EaseType.SineInOut);
        this.Loader[i].TweenFade(0,0.1).SetEase(FairyGUI.EaseType.QuadInOut).SetDelay(this._Interval - 0.1);
    }
    
    Update() {
        super.Update();
        this._Vy -= UpgradeShineParticleEffect._Gravity * TimeKit.DeltaTime;
        for (let i = 0; i < 2; i++) {
            this.Loader[i].x += this._Vx * TimeKit.DeltaTime;
            this.Loader[i].y += this._Vy * TimeKit.DeltaTime;
            this.Loader[i].SetScale(Mathf.RandomRange(0.7,1.3),Mathf.RandomRange(0.7,1.3));
            this.Loader[i].rotation = Mathf.Angle(this._Vx,this._Vy);
        }
        if(this.Loader[0].y <this._Floor){
            this._Vy = -this._Vy * 0.75;
            for (let i = 0; i < 2; i++) {
                this.Loader[i].y = this._Floor + 0.1;
            }
            this._Vx *= 1.1;
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/UpgradeShineParticleEffect.cs
using FairyGUI;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class UpgradeShineParticleEffect : AbstractEffect
    {
        public FairyGUI.GLoader[] mLoader = new FairyGUI.GLoader[2];
        private float mInterval;
        private float mScale;
        private float mVx;
        private float mVy;
        private float mFloor;
        private static int mGravity = 3000;

        public UpgradeShineParticleEffect(float x,float y)
        {
            for (int i = 0; i < 2; i++)
            {
                mLoader[i] = UIObjectFactory.NewObject(ObjectType.Loader).asLoader;
                mInterval = Random.Range(0.5f,1f);
                mLoader[i].SetPivot(0.5f,0.5f,true);
                mLoader[i].color = new Color32((byte)Random.Range(127,255),(byte)Random.Range(102, 153),(byte)Random.Range(0, 51),255);
                if(mInterval > 0.9)
                    mLoader[i].alpha = 0;
                else
                    mLoader[i].alpha = 1;
                
                mLoader[i].rotation = Random.Range(0,360);
                mVx = Random.Range(-900,900);
                mVx = Random.Range(-500,1000);
                mFloor = Random.Range(100,250);
                mLoader[i].blendMode = BlendMode.Add;
                mLoader[i].SetPosition(x,y,0);
                if(this.mInterval > 0.9)
                    this.mLoader[i].TweenFade(1,Mathf.Clamp(this.mInterval - 0.9f,0,0.1f)).SetEase(EaseType.SineInOut);
                mLoader[i].TweenFade(0,0.1f).SetEase(FairyGUI.EaseType.QuadInOut).SetDelay(this.mInterval - 0.1f);
            }
        }

        public override void Update()
        {
            base.Update();
            mVy -= mGravity * Time.deltaTime;
            for (var i = 0; i < 2; i++) {
                mLoader[i].x += mVx * Time.deltaTime;
                mLoader[i].y += mVy * Time.deltaTime;
                mLoader[i].SetScale(Random.Range(0.7f,1.3f),Random.Range(0.7f,1.3f));
                mLoader[i].rotation = Angle(mVx,mVy);
            }
            if(mLoader[0].y <mFloor){
                mVy = -mVy * 0.75f;
                for (var i = 0; i < 2; i++) {
                    mLoader[i].y = mFloor + 0.1f;
                }
                mVx *= 1.1f;
            }
        }
        
        private static float Angle(float x,float y){
            var angle = Mathf.Atan2(y,x) * Mathf.Rad2Deg;
            if(angle<0)
                angle += 360;
            return angle;
        }
    }
}